### PR TITLE
Save/load list of node

### LINKED ISSUE #274: Saving and loading for `Iterable[BaseNode]` objects

### LINKED ISSUE #287: bug in reachable
reachable has a bug:
```python 

backend='numpy'
nodes = [
  tn.Node(
      np.random.rand(2, 2, 2, 2),
      backend=backend,
      name='Node{}'.format(n),
      axis_names=[
          'node{}_1'.format(n), 'node_{}_2'.format(n),
          'node_{}_3'.format(n), 'node_{}_4'.format(n)
      ]) for n in range(4)
]
_ = [nodes[n][0] ^ nodes[n + 1][1] for n in range(3)]
subset = [nodes[0],nodes[1]]
print([b.name for b in tn.reachable(subset)]) #this should always print out all node names; however, sometimes it doesn't (you have to run it several times).
```
The problem is this:
```python 
def _reachable(nodes: Set[BaseNode]) -> Set[BaseNode]:
  if not nodes:
    raise ValueError("Reachable requires atleast 1 node.")
  node_que = collections.deque()
#==========bug===========
  seen_nodes = nodes #this should be an empty set
#============================
  node_que.append(list(nodes)[0])
  while node_que:
    node = node_que.popleft()
    for e in node.edges:
      for n in e.get_nodes():
        if n is not None and n not in seen_nodes:
          node_que.append(n)
          seen_nodes.add(n)
  return seen_nodes

```
*edit: it should be this
```python
def _reachable(nodes: Set[BaseNode]) -> Set[BaseNode]:
  if not nodes:
    raise ValueError("Reachable requires atleast 1 node.")
  node_que = collections.deque()
  initial_node= list(nodes)[0]
  seen_nodes = {initial_node}
  node_que.append(initial_node)
  while node_que:
    node = node_que.popleft()
    for e in node.edges:
      for n in e.get_nodes():
        if n is not None and n not in seen_nodes:
          node_que.append(n)
          seen_nodes.add(n)
  return seen_nodes

```

### PR DESCRIPTION
**Pull request by mganahl**
- Add support for saving and loading `List[BaseNode]`
- add `tn.contract_trace_edges(BaseNode)` 
- fix bug in reachable

fixes #287 
fixes #274

### CODE REVIEW

File: tensornetwork/tests/network_operations_test.py [hunk 1]
```diff
@@ -217,3 +217,12 @@ def test_reachable(backend):
...
   assert set(nodes) == tn.reachable(nodes[0])
+
+
+def test_reachable_disconnected(backend):
+  nodes = [tn.Node(np.random.rand(2, 2, 2), backend=backend) for _ in range(4)]
+  nodes[1][1] ^ nodes[2][0]  #connect 2nd and third node
+  assert set(tn.reachable([nodes[0],
+                           nodes[1]])) == {nodes[0], nodes[1], nodes[2]}
+  nodes[2][1] ^ nodes[3][0]  #connect third and fourth node
+  assert set(tn.reachable([nodes[0], nodes[1]])) == set(nodes)
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_operations_test.py]: Can you add another test where the 0th node is connected to something?

So test a network that looks like

A-B
C-D

File: tensornetwork/utils.py [hunk 2]
```diff
@@ -1,28 +1,150 @@
...
+        'node{}'.format(n): v
+        for n, v in enumerate(net_file["node_names"]['names'][()])
+    }
+
+    edge_names = {
+        'edge{}'.format(n): v
+        for n, v in enumerate(net_file["edge_names"]['names'][()])
+    }
+    edges = list(net_file["edges"].keys())
+    #print(net_file["nodes"]['node_names'][()])
```
Comment (Thenerdstation) [inline tensornetwork/utils.py]: Delete commented out code. (we really should have pylint check for this).

File: tensornetwork/network_operations.py [hunk 3]
```diff
@@ -549,8 +549,9 @@ def _reachable(nodes: Set[BaseNode]) -> Set[BaseNode]:
   if not nodes:
     raise ValueError("Reachable requires atleast 1 node.")
   node_que = collections.deque()
-  seen_nodes = nodes
-  node_que.append(list(nodes)[0])
+  initial_node = list(nodes)[0]
```
[the remark this answers is not in the document]
Comment (mganahl) [reply tensornetwork/network_operations.py]: added

File: tensornetwork/tests/network_components_free_test.py [hunk 4]
```diff
@@ -802,3 +802,63 @@ def test_broken_edge_contraction_magicmethod(backend):
...
+          backend=backend,
+          name='Node{}'.format(n),
+          axis_names=[
+              'node{}_1'.format(n), 'node_{}_2'.format(n),
+              'node_{}_3'.format(n), 'node_{}_4'.format(n)
+          ]) for n in range(4)
+  ]
+  _ = [nodes[n][0] ^ nodes[n + 1][1] for n in range(3)]
+  with pytest.raises(ValueError):
+    tn.save_nodes([nodes[0], nodes[1]], 'test_file_save_nodes')
```
[the remark this answers is not in the document]
Comment (mganahl) [reply tensornetwork/tests/network_components_free_test.py]: fixed

File: tensornetwork/tests/network_operations_test.py (hunk 1 above)
Comment (mganahl) [reply tensornetwork/tests/network_operations_test.py]: added

File: tensornetwork/utils.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/utils.py]: removed

File: tensornetwork/tests/network_operations_test.py [hunk 5]
```diff
@@ -217,3 +217,36 @@ def test_reachable(backend):
...
+
+  assert set(tn.reachable([nodes[0]])) == {nodes[0], nodes[1]}
+  assert set(tn.reachable([nodes[1]])) == {nodes[0], nodes[1]}
+  assert set(tn.reachable([nodes[0], nodes[1]])) == {nodes[0], nodes[1]}
+
+  assert set(tn.reachable([nodes[2]])) == {nodes[2], nodes[3]}
+  assert set(tn.reachable([nodes[3]])) == {nodes[2], nodes[3]}
+  assert set(tn.reachable([nodes[2], nodes[3]])) == {nodes[2], nodes[3]}
+
+  assert set(tn.reachable([nodes[0], nodes[1],
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_operations_test.py]: Why does reachable with `nodes[0]` not also return `nodes[0]`?

Comment (Thenerdstation) [reply tensornetwork/tests/network_operations_test.py]: Based on this failing test, it seems like the answer is "It does"

```
 assert set(tn.reachable([nodes[0], nodes[1],
                               nodes[2]])) == {nodes[2], nodes[3]}
E     assert {<tensornetwo...7f7120f1c4a8>} == {<tensornetwor...7f7120f1c4a8>}
E       Extra items in the left set:
E       <tensornetwork.network_components.Node object at 0x7f7120f03cc0>
E       <tensornetwork.network_components.Node object at 0x7f7120f03668>
E       Use -v to get the full diff
```

Comment (mganahl) [reply tensornetwork/tests/network_operations_test.py]: thanks, sorry hadn't fixd them yet

**Pull request merged**